flink-runtime/src/main/java/org/apache/flink/runtime/metrics/dump/MetricDumpSerialization.java

```diff
@@ -124,55 +160,135 @@ public MetricSerializationResult serialize(
+							" so maybe not all metrics would be reported.");
+						markUnserializedMetrics = true;
+						//clear all, because we can not revoke the latest metrics which caused overflow
+						unserializedCounters.put(entry.getKey(), entry.getValue());
+						countersBuffer.clear();
+						numCounters = 0;
+					}
 				} catch (Exception e) {
 					LOG.debug("Failed to serialize counter.", e);
 				}
 			}
 
+			gaugesBuffer.clear();
 			int numGauges = 0;
 			for (Map.Entry<Gauge<?>, Tuple2<QueryScopeInfo, String>> entry : gauges.entrySet()) {
+				if (markUnserializedMetrics) {
+					unserializedGauges.put(entry.getKey(), entry.getValue());
+					continue;
+				}
+
 				try {
-					serializeGauge(buffer, entry.getValue().f0, entry.getValue().f1, entry.getKey());
+					serializeGauge(gaugesBuffer, entry.getValue().f0, entry.getValue().f1, entry.getKey());
 					numGauges++;
+					if (gaugesBuffer.length() + countersBuffer.length() > maximumFramesize) {
+						LOG.warn("The serialized gauge metric is larger than the maximum frame size, " +
+							" so maybe not all metrics would be reported.");
+						markUnserializedMetrics = true;
+						unserializedGauges.put(entry.getKey(), entry.getValue());
+						gaugesBuffer.clear();
+						numGauges = 0;
+					}
 				} catch (Exception e) {
 					LOG.debug("Failed to serialize gauge.", e);
 				}
 			}
 
+			histogramsBuffer.clear();
 			int numHistograms = 0;
 			for (Map.Entry<Histogram, Tuple2<QueryScopeInfo, String>> entry : histograms.entrySet()) {
+				if (markUnserializedMetrics) {
+					unserializedHistograms.put(entry.getKey(), entry.getValue());
+					continue;
+				}
+
 				try {
-					serializeHistogram(buffer, entry.getValue().f0, entry.getValue().f1, entry.getKey());
+					serializeHistogram(histogramsBuffer, entry.getValue().f0, entry.getValue().f1, entry.getKey());
 					numHistograms++;
+					if (histogramsBuffer.length() + gaugesBuffer.length() + countersBuffer.length() > maximumFramesize) {
+						LOG.warn("The serialized histogram metric is larger than the maximum frame size, " +
+							" so maybe not all metrics would be reported.");
+						markUnserializedMetrics = true;
+						unserializedHistograms.put(entry.getKey(), entry.getValue());
+						histogramsBuffer.clear();
+						numHistograms = 0;
+					}
 				} catch (Exception e) {
 					LOG.debug("Failed to serialize histogram.", e);
 				}
 			}
 
+			metersBuffer.clear();
 			int numMeters = 0;
 			for (Map.Entry<Meter, Tuple2<QueryScopeInfo, String>> entry : meters.entrySet()) {
+				if (markUnserializedMetrics) {
+					unserializedMeters.put(entry.getKey(), entry.getValue());
+					continue;
+				}
+
 				try {
-					serializeMeter(buffer, entry.getValue().f0, entry.getValue().f1, entry.getKey());
+					serializeMeter(metersBuffer, entry.getValue().f0, entry.getValue().f1, entry.getKey());
 					numMeters++;
+					if (metersBuffer.length() + histogramsBuffer.length() + gaugesBuffer.length() +
```

Comment:
this implies that histograms are prioritized over meters which is not correct IMO.

Reply:
Hi, @zentol I don't quite understand what this means?

Reply:
Let's say that our max frameSize is 100 bytes. Let's also say that counters, meters and gauges take up 1 byte, and histograms take up 98.

I the order that you're serializing metrics (counters -> gauges -> histograms -> meters), counters, gauges and histograms will be serialized and fit (1 + 1 + 98 <= 100). Then the meters come along with their additional 1 byte and are now dropped from the report.

Given that the only known case of this problem arising is a ridiculous amount of latency histograms they should be excluded first, instead of meters. As it stands histograms are the least important yet most expensive metrics, so let's drop those first instead of rather essential metrics like `numRecordsInPerSecond`.

Now, this does _not_ mean that you now should change the order again (like you had in a previous version), but that the MQS should first try to drop histograms to fit the report into the dump.